File: huffman.js
//Input string as command line argument

function frequency(str){
	var freqs={};
	for (var i in str){
		if (freqs[str[i]]==undefined){
			freqs[str[i]]=1;
		}
		else{
			freqs[str[i]]+=1;
		}
	}
	return freqs;
}
function sort(freq){
	chars=[];
	for (i in freq){
		chars.push([freq[i],i]);
	}
	return (chars.sort());
}
function buildTree(list){
	while ((list.length)>1){
		var firstTwo=list.slice(0,2);
		var rest=list.slice(2);
		var add=firstTwo[0][0]+firstTwo[1][0];
		list=rest;
		var addedList=[add,firstTwo];
		list.push(addedList);
		list.sort();
	}
	return list[0];
}
function trimTree(tree){
	var p=tree[1];
	if (typeof(p)==typeof("")){
		return p;
	}
	else{
		return (Array(trimTree(p[0]),trimTree(p[1])));
}
}
var codes={};
function assignCodes(node,pat){
	pat=pat || "";
	if (typeof(node)==typeof("")){
		codes[node]=pat;
	}
	else{
		assignCodes(node[0],pat+"0");
		assignCodes(node[1],pat+"1");
	}
}
function encode(str){
	output="";
	for (var letter in str){
		output+=codes[str[letter]];
	}
	return output;
}
function decode(treee,bitstr){
	var out="";
	var tree=treee;
	for (var bit in bitstr){
		if (bitstr[bit]=="0"){
			tree=tree[0];
		}
		else{
			tree=tree[1];
		}
		if (type(tree)==type("")){
			out+=tree;
			tree=treee
		}
	}
	return out;
}
assignCodes(trimTree(buildTree(sort(frequency(process.argv[2])))));
console.log("To encode: "+process.argv[2]);
console.log("Tree: ",trimTree(buildTree(sort(frequency(process.argv[2])))));
console.log("Binary Codes for each character: "+codes);
console.log("Encoded: "+encode(process.argv[2]));
console.log("Decoded: "+decode(trimTree(buildTree(sort(frequency(process.argv[2]))))),process.argv[2]);
console.log("Here "+process.argv[2]+" need only "+encode(process.argv[2]).length+" bits");

File: huffman.py
# Input string as command line argument

import sys

codes={}
def frequency(strin):
	freq={}
	for i in strin:
		freq[i]=freq.get(i,0)+1
	return sort(freq)
def sort(freq):
	tuples=[]
	for i in freq.keys():
		tuples.append((freq[i],i))
		tuples.sort()
	return buildtree(tuples)
def buildtree(tuples):
	while len(tuples)>1:
		first_two=tuple(tuples[:2])
		rest=tuples[2:]
		add=first_two[0][0]+first_two[1][0]
		tuples=rest+[(add,first_two)]
		tuples.sort()
	return trimtree(tuples[0])
def trimtree(tree):
	p=tree[1]
	if type(p)==type(""):
		return p
	else:
		return (trimtree(p[0]),trimtree(p[1]))
def assigncode(node,pattern=""):
	global codes
	if type(node)==type(""):
		codes[node]=pattern
	else:
		assigncode(node[0],pattern+"0")
		assigncode(node[1],pattern+"1")
def encode(string):
	global codes
	output=""
	for i in string:
		output +=codes[i]
	return output
def decode(treee,bitstr):
	out=""
	tree=treee
	for bit in bitstr:
		if bit=="0":tree=tree[0]
		else:tree=tree[1]
		if type(tree)==type(""):
			out+=tree
			tree=treee
	return out
	
assigncode(frequency(sys.argv[1]))
print "To encode:",sys.argv[1]
print "character bits:",codes
print "encoded bits:",encode(sys.argv[1])
print "tree=",frequency(sys.argv[1])
print "decoded:",decode(frequency(sys.argv[1]),encode(sys.argv[1]))
print "Here \"%s\" need only %d bits where generally needs %d bits"%(sys.argv[1],len(encode(sys.argv[1])),8*len(sys.argv[1]))
